src/tracers.jl

```diff
@@ -1,194 +1,184 @@
 abstract type AbstractTracer <: Number end
 
+# We represent sparse Hessians as sets of index tuples.
+const AbstractPairSet{T} = AbstractSet{Tuple{T,T}}
+
 # Convenience constructor for empty tracers
 empty(tracer::T) where {T<:AbstractTracer} = empty(T)
+empty(T) = T()
 
-const SET_TYPE_MESSAGE = """
-The provided index set type `S` has to satisfy the following conditions:
+sparse_vector(T, index) = T([index])
 
-- it is an iterable with `<:Integer` element type
-- it implements `union`
+#================#
+# Elementwise OR #
+#================#
 
-Subtypes of `AbstractSet{<:Integer}` are a natural choice, like `BitSet` or `Set{UInt64}`.
-"""
+×(a::G, b::G) where {G<:AbstractSet} = Set((i, j) for i in a, j in b)
```

Comment:
what about
```julia
Set(Base.Iterators.product(a, b))
```

Reply:
Is that more performant / type stable?

Otherwise I find the verbose option a bit more readable. Afaik, it calls `Base.Iterators.product` as well.

Reply:
then let's keep it, I have no specific argument for either